### PR TITLE
use ref count for participant workers

### PR DESCRIPTION
**Pull request by paulwe**

### CODE REVIEW

File: pkg/rtc/room.go [hunk 1]
```diff
@@ -575,14 +575,15 @@ func (r *Room) RemoveParticipant(identity livekit.ParticipantIdentity, pID livek
 		return
 	}
 
+	r.removeParticipantWorkerLocked(p)
```
Comment (paulwe) [inline pkg/rtc/room.go:578]: we can't skip cleaning this up if the new participant is added before the old one is removed or we'll leak the worker

File: pkg/rtc/room.go [hunk 2]
```diff
@@ -1525,15 +1517,7 @@ func (r *Room) addParticipantWorkerLocked(p types.LocalParticipant) *participant
 func (r *Room) removeParticipantWorkerLocked(p types.LocalParticipant) {
 	identity := p.Identity()
 	if pw, ok := r.participantWorkers[identity]; ok {
-		n := len(pw.participants)
-		for idx, participant := range pw.participants {
-			if p == participant {
-				pw.participants[idx] = pw.participants[n-1]
-				pw.participants = pw.participants[:n-1]
```
Comment (paulwe) [inline pkg/rtc/room.go:1532]: to avoid a leak here we would want to set `pw.participants[n-1]` to nil but i think we can just use a ref count here?

**Review by paulwe — COMMENTED** [review]
a couple of things i didn't notice when i read the pr

**Comment by boks1971** [comment]
I had ref count initially @paulwe , but then I realised that remove may be called more than once. I was not entirely sure if all paths worked properly. I think it should, but just went the safe route.

File: pkg/rtc/room.go (hunk 1 above)
Comment (boks1971) [reply pkg/rtc/room.go:578]: good catch, missed that.

**Comment by paulwe** [comment]
@boks1971 that's fine. skipping the remove is actually a bug though right?

File: pkg/rtc/room.go (hunk 2 above)
Comment (boks1971) [reply pkg/rtc/room.go:1532]: yeesh, missed that too, good catch once more!

**Comment by boks1971** [comment]
> @boks1971 that's fine. skipping the remove is actually a bug though right?

Yes it is. And not setting to `nil` is also bad. Two great catches.

**Comment by boks1971** [comment]
I will make a PR with those two things changed.

**Pull request closed without merging**